File: src/main/java/NeuralNetwork/NeuralNetwork.java
package NeuralNetwork;

import NeuralNetwork.exceptions.NoHiddenLayersException;

import java.util.ArrayList;
import java.util.stream.IntStream;

public class NeuralNetwork {

    private Float[][] inputs;
    private Integer[] outputs;

    private int numOfInputs;

    private float calcForwardOutput;
    private ResultParser resultParser;
    private Propagator propagator;
    private ArrayList<HiddenLayer> hiddenLayers;

    public NeuralNetwork(Float[][] inputs, Integer[] outputs){
        this.inputs = inputs;
        this.outputs = outputs;
        this.resultParser = new ResultParser();
        this.numOfInputs = inputs[0].length;
        hiddenLayers = new ArrayList<>();
    }

    public void addHiddenLayer(int neurons){
        HiddenLayer hiddenLayer = new HiddenLayer(neurons, this.numOfInputs);
        hiddenLayers.add(hiddenLayer);
    }

    /**
     * Creates a single hidden layer then executes forward and backward propagation
     * @param iterations   Number of times iterated through the network
     * @throws Exception
     */
    public void train(int iterations) throws Exception {
        if (inputs.length != outputs.length)
            throw new Exception(String.format("Inputs and Outputs are not identical. Input Count: %d, Output Count: %d", inputs.length, outputs.length));
        if(hiddenLayers.size() == 0)
            throw new NoHiddenLayersException("Need to add hidden layers!");

        System.out.println("Start process...");
        System.out.println(String.format("Number of Iterations: %d", iterations));
        resultParser.resetNSuccess();
        IntStream.range(0, iterations).forEach(i ->
                IntStream.range(0, outputs.length).forEach(j -> {
                    propagator = new Propagator(hiddenLayers.get(0).getInputWeights(), hiddenLayers.get(0).getOutputWeights(), hiddenLayers.get(0).getNeurons());
                    calcForwardOutput = propagator.forward(inputs[j].clone());
                    propagator.backward(outputs[j], hiddenLayers.get(0).getInputWeights(), hiddenLayers.get(0).getOutputWeights(), hiddenLayers.get(0).getNeurons(), inputs[j].clone());
                    hiddenLayers.get(0).setInputWeights(propagator.getInputWeights());
                    hiddenLayers.get(0).setOutputWeights(propagator.getOutputWeights());
                    resultParser.countSuccess(calcForwardOutput, outputs[j]);
                }));
        final float accuracy = (resultParser.getNSuccess() / (float)(iterations * outputs.length)) * 100;
        System.out.println(String.format("Accuracy: %s", accuracy));
    }

    public int predict(Float[] input){
        return resultParser.parseResult(propagator.forward(input));
    }

}

File: src/main/kotlin/neuralNetwork/NeuralNetwork.kt
package neuralNetwork

import neuralNetwork.exceptions.NoHiddenLayersException
import java.util.concurrent.ThreadLocalRandom
import java.util.stream.IntStream

/**
 * Creates the weights from inputs-to-hidden and hidden-to-outputs
 */
class HiddenLayer(val neurons: Int, inputs: Int) {
    var outputWeights = Array(neurons) { ThreadLocalRandom.current().nextFloat() }
    var inputWeights = Array(inputs) { FloatArray(neurons) { ThreadLocalRandom.current().nextFloat() } }
}

class NeuralNetwork(private val inputs: Array<Array<Float>>, private val outputs: Array<Int>) {

    private val resultHandler: ResultHandler = ResultHandler()
    private val hiddenLayers: ArrayList<HiddenLayer> = ArrayList()
    private var propagator: Propagator? = null

    fun addHiddenLayer(neurons: Int) {
        val hiddenLayer = HiddenLayer(neurons, inputs[0].size)
        hiddenLayers.add(hiddenLayer)
    }

    /**
     * Creates a single hidden layer then executes forward and backward propagation
     * @param iterations   Number of times iterated through the network
     * @throws Exception
     */
    fun train(iterations: Int) {
        if (inputs.size != outputs.size)
            throw Exception("Inputs and Outputs are not identical. Input Count: ${inputs.size}, Output Count: ${outputs.size}")
        if (hiddenLayers.size == 0)
            throw NoHiddenLayersException("Need to add hidden layers!")

        println("Start process...")
        println("Number of Iterations: $iterations")
        resultHandler.resetNSuccess()
        for (i in 0 until iterations) {
            IntStream.range(0, outputs.size).forEach { j ->
                val tempPropagator =
                    Propagator(hiddenLayers[0].inputWeights, hiddenLayers[0].outputWeights, hiddenLayers[0].neurons)
                val calcForwardOutput = tempPropagator.forward(inputs[j].clone())
                tempPropagator.backward(
                    outputs[j].toFloat(),
                    hiddenLayers[0].inputWeights,
                    hiddenLayers[0].outputWeights,
                    hiddenLayers[0].neurons,
                    inputs[j].clone()
                )
                hiddenLayers[0].inputWeights = tempPropagator.inputWeights
                hiddenLayers[0].outputWeights = tempPropagator.outputWeights
                resultHandler.countSuccess(calcForwardOutput, outputs[j].toFloat())
                propagator = tempPropagator
            }
        }
        val accuracy = resultHandler.nSuccess / (iterations * outputs.size).toFloat() * 100
        println("Accuracy: $accuracy")
    }

    fun predict(input: Array<Float>): Int {
        return resultHandler.parseResult(propagator!!.forward(input))
    }

}
